### DISCUSSION TITLE
Proposal for Apache Incubator

### DISCUSSION THREAD

**Discussion by acezen** [apache/incubator-graphar#321] [General]
Greetings!

We are proposing to enter GraphAr, open source and language-independent data file format designed for efficient graph data storage and retrieval, into incubation. Please see the proposal below and let us know if you have any questions or concerns.

## Abstract

GraphAr is an open source and language-independent data file format designed for efficient graph data storage and retrieval.

## Proposal

GraphAr provides a standard data file format for graph data storage and exchange with the following features: 

- **Efficient format design**:
  - **Chunk-based**: GraphAr utilizes a chunk-based design, partitioning the graph data into chunks, with each stored in a separate file. This design allows for easy, parallelized access and distribution of the graph data across different machines.
  - **Columnar storage**: GraphAr leverages high-performance columnar storage formats, including Parquet and ORC, for organizing graph data files, which allows for optimized access patterns.
  - **Metadata management**: GraphAr maintains graph metadata using a set of YAML files, simplifying the process of understanding and utilizing the schema information.
  - **Maintain CSR/CSC semantics**: By sorting edges by source vertex id and destination vertex id, GraphAr maintains the CSR/CSC semantics of graph data. This eliminates the need to reconstruct the CSR/CSC structure when loading the graph data.
- **Out-of-core queries**: GraphAr is designed for out-of-core scenarios, enabling the storage and querying of large-scale graphs outside of memory, such as in data lakes.
- **Cross-language support**: GraphAr provides libraries in C++, Java, Scalar with Spark and Python with PySpark for generating, accessing, and transforming files in GraphAr format.

GraphAr currently supports cross-language operations, including C++, Java, and Spark, offering performance up to 6 times faster than CSV in loading graph data. It also facilitates the exchange of graph data between different graph systems, such as Nebula Graph, HugeGraph, and Neo4j.

## Background

In recent years, several graph store systems, such as vineyard, groot, and GART, have been integrated into the GraphScope system. Each system has its own graph data storage layout, complicating the exchange of graph data between different systems. Moreover, there is no standard format for large graph data storage and exchange in the graph community.

To address this gap, we decided to develop GraphAr.

Initially, the project was developed by the Alibaba GraphScope team in 2022 and was open-sourced on GitHub in December 2022. A number of systems at Alibaba, including GraphScope and vineyard, have adopted GraphAr as their standard graph data storage format.

## Rationale

Numerous graph systems, such as Neo4j, Nebula Graph, and HugeGraph, have been developed in recent years. Each of these systems has its own graph data storage format, complicating the exchange of graph data between different systems. The need for a standard data file format for large-scale graph data storage and processing that can be used by diverse existing systems is evident, as it would reduce overhead when various systems work together.

Our aim is to fill this gap and contribute to the open-source community by providing a standard data file format for graph data storage and exchange, as well as for out-of-core querying. This format, which we have named GraphAr, is engineered to be efficient, cross-language compatible, and to support out-of-core processing scenarios, such as those commonly found in data lakes. Furthermore, GraphAr's flexible design ensures that it can be easily extended to accommodate a broader array of graph data storage and exchange use cases in the future.

## Initial Goals

  * Build a more diverse community with contributors from different organizations.
  * Facilitate the adoption and integration of GraphAr by ensuring its neutrality and durability.
  * Collect feedback from the community to standardize the format and improve its efficiency.

### Meritocracy

This proposal intends to start build a community around GraphAr following the ASF meritocracy model. 
Users and new contributors will be respected and welcomed.
They will earn credit by participating in the community and providing quality contributions to move the project forward,the contributions not only include the code contributions but also the non-code contributions(documentation, discussion, testing, events, community development, etc.)
Those who make long-term and high quality contributions will be encouraged to become committers.

### Community

GraphAr is being developed by the development team inside Alibaba Group who's responsible for building Graph processing system too.
The extensive graph community is in need of a standardized file format. By incorporating GraphAr into the Apache ecosystem, we anticipate further expansion of the community, benefiting from increased collaboration and adoption of a common format.

### Users

GraphAr currently serves a group of users, including Alibaba, Fabarta and TuGraph. Here are some use cases of GraphAr:
	
  - Alibaba [GraphScope](https://github.com/alibaba/GraphScope) use GraphAr as graph data archive format [1].
  - In-memory data manager [vineyard](https://github.com/v6d-io/v6d) use GraphAr as graph object archive format.
  - [Fabarta](https://www.fabarta.com/) use GraphAr as their graph data archive and source to load. 
  - [TuGraph](https://github.com/TuGraph-family/tugraph-db) is considering using GraphAr as data format to exchange graph data with neo4j, and they have a [WIP Pull Request](https://github.com/TuGraph-family/tugraph-db/pull/277) to add GraphAr support. 

### Developer

As for developers, although GraphAr has attracted 13 contributors since it was open-sourced, we have four core developers: Weibin Zeng, Xue Li, Zhe Wang and Semyon Sinchenko. Weibin Zeng and Xue Li are the founders of GraphAr project, Zhe Wang is the author and maintainer of GraphAr JAVA module and Sem is the author and maintainer of GraphAr PySpark module.

The community's size and diversity are indeed areas of concern. However, we expect to attract more contributors in the future to address these issues by evolving the software and adhering to the Apache way.

The need for a standard graph data format is common in the graph community, and it provides the potential for a bigger community.

### Core Developers

   * Weibin Zeng (GitHub ID: [acezen](https://github.com/acezen)): Works in Alibaba Group, GraphScope, vineyard project committer.
   * Xue Li: (GitHub ID: [lixueclaire](https://github.com/lixueclaire)): Works in Alibaba Group, GraphScope committer.
   * Zhe Wang: (GitHub ID: [Thespica](https://github.com/Thespica)): The author of GraphAr Java module, an open-source enthusiast from Southwest Minzu University.
   * Semyon Sinchenko: (Github ID: [SemyonSinchenko](https://github.com/SemyonSinchenko)):  The author of GraphAr-PySpark module, an individual open-source enthusiast from  Raiffeisen Bank International.

### Project Name

We have checked and believe that GraphAr is an appropriate and easy-to-remember name.

### Orphaned Products

GraphAr is used as a graph data archive format in Alibaba's graph system GraphScope and vineyard. The developers of these systems will continue to improve GraphAr to meet current and future requirements. Other organizations, such as Fabarta, also use GraphAr in their core products. Furthermore, TuGraph and HugeGraph has shown interest in GraphAr and is considering using it as their graph data archive format. Given the extensive need in the graph community for a standardized file format, we believe the developer and user communities will continue to grow, mitigating the risk of GraphAr becoming an orphaned product.

### Inexperience with Open Source

The creators of GraphAr have been working on open-source projects for many years. They has been dedicate to open source projects, such as GraphScope and vineyard. 

### Homogenous Developers

Currently, GraphAr has four core developers. However, we are working to diversify our developer base. We have already received interest from other organizations, such as Fabarta, which are using GraphAr in their core products. We believe that the developer and user communities will continue to grow.

### Reliance on Salaried Developers

It is true that most developers are supported by their employers to contribute to GraphAr, which poses a significant risk. However, GraphAr has already been used by GraphScope and deployed within Alibaba, with no internal forked versions. The developers of GraphScope will continue to improve GraphAr to meet current and future requirements. As a result, Alibaba can ensure a long-term commitment. We believe that if GraphAr enters the incubator, we can attract more maintainers and developers from diverse backgrounds to address this risk.

### Relationships with Other Apache Products

GraphAr relies on Parquet, ORC, and Arrow for data storage and exchange and on Spark to provide a connector to graph databases like Nebula Graph, HugeGraph, and Neo4j.

GraphAr can also be integrated with several Apache projects for graph data storage and exchange, including:
  - [HugeGraph](https://hugegraph.apache.org): A convenient, efficient, and adaptable graph database.
  - [AGE](https://age.incubator.apache.org/): Graph database optimized for fast analysis and real-time data processing
  - [TinkerPop](https://tinkerpop.apache.org/): A graph computing framework for both graph databases (OLTP) and graph analytic systems (OLAP).

We believe that such integration could also be applied to and benefit the ope-source community, and we have a plan to discuss with the community to make it happen.

### An Excessive Fascination with the Apache Brand

We believe that the Apache way and its neutrality, not just the brand, will help GraphAr grow. The need for a standard graph data format is common in the graph community and is relevant to many other graph system projects, not just those in Alibaba. A neutral organization like Apache will ultimately better serve the community than a single company.

## Documentation

GraphAr documentation is provided on [docs](https://alibaba.github.io/GraphAr/).

## Initial Source

GraphAr has been under development since June 2022 by a team of engineers at Alibaba called GraphScope. It was open-sourced on GitHub in December 2022, with the project available at https://github.com/alibaba/GraphAr under the name GraphAr. The project is licensed under Apache License 2.0.

## Source and Intellectual Property Submission Plan

As soon as GraphAr is approved to join Apache Incubator, our initial committers will submit iCLA(s), SGA, and CCLA(s). The codebase is already licensed under Apache License 2.0.

We will also deprecate the initial source repository and redirect it to the new incubator project repository after approved.

## External Dependencies

GraphAr has several external dependencies with various licenses, including Apache 2.0, BSD, BSL-1.0, and MIT.

 - Apache 2.0
   - https://github.com/apache/arrow
   - OpenSSL
   - https://github.com/alibaba/fastFFI
   - https://github.com/google/benchmark
   - com.aliyun.odps:hadoop-fs-oss:jar:3.3.8-public
   - com.aliyun.odps:odps-spark-datasource_2.11:jar:3.3.8-public
   - com.aliyun.odps:cupid-sdk:jar:3.3.8-public	
   - com.vesoft:nebula-spark-connector_3.0:jar:3.6.0
   - org.neo4j:neo4j-connector-apache-spark_2.12:jar:5.0.0_for_spark_3
   - org.scala-lang.modules:scala-collection-compat_2.12:jar:2.1.1
   - org.yaml:snakeyaml:jar:1.26
   - org.apache.spark:spark-core_2.12:jar:3.2.2
   - org.apache.spark:spark-mllib_2.12:jar:3.2.2
   - org.apache.spark:spark-sql_2.12:jar:3.2.2
   - org.apache.spark:spark-streaming_2.12:jar:3.2.2
   - org.scala-lang:scala-library:jar:2.12.10
   - org.scalatest:scalatest_2.12:jar:3.1.1
   - com.alibaba.fastffi:annotation-processor:jar:0.1.2
   - com.alibaba.fastffi:ffi:jar:0.1.2
   - com.alibaba.fastffi:llvm:jar:0.1.2
   - com.alibaba.fastffi:llvm4jni-runtime:jar:0.1.2
   - org.apache.arrow:arrow-c-data:jar:13.0.0
   - org.apache.arrow:arrow-dataset:jar:13.0.0
   - org.apache.arrow:arrow-memory-netty:jar:13.0.0
   - org.apache.arrow:arrow-vector:jar:13.0.0
   - https://github.com/google/styleguide/blob/gh-pages/cpplint/cpplint.py
   - pyspark

 - BSD
   - junit:junit:jar:4.13.2
   - Python

 - BSL-1.0
   - Catch2
   - Boost
     
 - MIT
   - https://github.com/bitwizeshift/result
   - https://github.com/jimmiebergmann/mini-yaml
   - org.slf4j:slf4j-simple:jar:1.7.25
   - pytest
   - pytest-cov
   - pyyaml
   - ruff

## Cryptography

N/A

### Mailing lists

private@graphar.incubator.apache.org
dev@graphar.incubator.apache.org
commits@graphar.incubator.apache.org

### Subversion Directory
N/A

### Git Repositories

Upon entering incubation, we want to transfer the existing repo to the Apache Software Foundation:
-  https://github.com/alibaba/GraphAr to https://github.com/apache/incubator-graphar
-  https://github.com/GraphScope/gar-test to https://github.com/apache/graphar-testing

### Issue Tracking

The community would like to continue using GitHub Issues.

### Other Resources

The community has chosen GitHub actions as its continuous integration tools.

## Initial Committers

* Weibin Zeng (Alibaba, weibin.zen@gmail.com)
* Xue Li (Alibaba, lx_claire@qq.com)
* Zhe Wang (Southwest Minzu University, thespica@qq.com)
* Semyon Sinchenko (Raiffeisen Bank International, ssinchenko@protonmail.com)
* He Tao (Alibaba, sighingnow@gmail.com)

### Champion
- Yu Li (liyu@apache.org)

### Nominated Mentors
- Calvin Kirs (kirs@apache.org)
- tison (tison@apache.org)
- Xiaoqiao He (hexiaoqiao@apache.org)
- Yu Li (liyu@apache.org)

### Sponsoring Entity

We are requesting the Incubator to sponsor this project.

## References

[1] https://graphscope.io/blog/tech/2023/08/31/Getting-Started-with-GraphAr-Standardized-Graph-Storage-File-Format

**Comment by Thespica** [comment]
LGTM, but there is a little typo:  In Initial Goals part, the second point, should be `its` rather than `tis`.

**Comment by SemyonSinchenko** [comment]
LGTM
P.S. Maybe add something about python in this line `Cross-language support: GAR provides libraries in C++ and Java for generating, accessing, and transforming files in GAR format.`? Like `C++, Java and Python with PySpark ...`? The popularity of python is growing very fast in both Network Analysis and Data Engineering fields.

**Reply by acezen** [reply]
good suggestion，updated.

**Comment by acezen** [comment]
hi, @SemyonSinchenko, can you help to list the external dependencies and license that introduced by GraphAr-PySpark here

**Reply by SemyonSinchenko** [reply]
- Python itself, PSFL (BSD-like)
- pyspark, Apache-2.0
- pytest, MIT
- pytest-cov, MIT
- pyyaml, MIT
- ruff, MIT

**Comment by acezen** [comment]
### PERFORM A SUITABLE NAME SEARCH
#### reference:
 - [Apache project naming policy](https://apache.org/foundation/marks/pmcs.html#naming)
 - [name process](https://www.apache.org/foundation/marks/naming.html)

#### Open Source Adoption
- Github: https://github.com/search?q=graphar&type=repositories&ref=simplesearch
 Notable matches:
 https://github.com/Joshuayangg/GraphAR  An iOS mobile AR application
 https://github.com/dromaniuk/graphar

- SF.net
  No Results
  https://sourceforge.net/directory/linux/?q=GraphAr
- Google Code
  No Results
  https://opensource.google/s/results?q=GraphAr
- openhub
  No Results
  https://openhub.net/p?ref=homepage&query=GraphAr

#### Use On Word Wide Web
- Google
 https://www.google.com/search?q=GraphAr
 Top result point to the project GraphAr
- Bing
  https://cn.bing.com/search?q=GraphAr&ensearch=1&FORM=BESBTB
  Top result point to the project GraphAr
- Yahoo
  https://search.yahoo.com/search?p=GraphAr
  Top result point to the project GraphAr

#### Registration
 - USPTO
   No Results
   https://search.uspto.gov/search?query=GraphAr&op=Search&affiliate=web-sdmg-uspto.gov
- Trademarkia
  No Results
   https://www.trademarkia.com/trademarks-search.aspx?tn=GraphAr
- Euipo
  No Results
  https://euipo.europa.eu/eSearch/#basic/1+1+1+1/GraphAr
